---
title: "QTL Mapping in Multi-Parent Populations"
author: "Wenhao Li, Martin Boer, and Bart-Jan van Rossum"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_vignette:
    toc: false
    number_sections: false
bibliography: bibliography.bib
link-citations: yes
vignette: >
  %\VignetteIndexEntry{QTL Mapping in Multi-Parent Populations}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.dim = c(7, 4)
)
library(statgenMPP)
op <- options(width = 90, 
              digits = 3)
```

## The statgenMPP package

The `statgenMPP` package is developed as an easy-to-use package for QTL mapping in  multi-parent populations. The package has many ways of visualizing inputs and results.

This vignette describes in detail how to perform the IBD calculations and how to do QTL mapping using the IBD probabilities in a mixed model framework, using the [LMMsolver](https://biometris.github.io/LMMsolver) R package. This will be done using two example data sets. The first data set is a barley data set for awn length described in @Liller2017-jz. The second data set is a maize data set, the dent panel of the EU-NAM maize project (@Giraud2014) 

## IBD Calculations

The calculations of the IBDs are based on Hidden Markov Models (HMM) and inheritance vectors and are performed using the [statgenIBD](https://biometris.github.io/statgenIBD) R package. For details of the theory see @Lander1987 and @Huang2011. It is also possible to import IBD probabilities computed using the [RABBIT](https://github.com/chaozhi/RABBIT) software [@Zheng2014; @Zheng2015; @Zheng2018].

### Population types

In the `statgenIBD` package, IBD probabilities can be calculated for many different types of populations. In the following table all supported populations are listed. Note that the value of x in the population types is variable, with its maximum value depicted in the last column. Using RABBIT more complex populations such as MAGIC can also be analyzed.

+---------------------+------------+------------------------------------------------------------+------------+
| **Population type** | **Cross**  | **Description**                                            | **max. x** |
+=====================+============+============================================================+============+
| DH                  | biparental | doubled haploid population                                 |            |
+---------------------+------------+------------------------------------------------------------+------------+
| Fx                  | biparental | Fx population (F1, followed by x-1 generations of selfing) | 8          |
+---------------------+------------+------------------------------------------------------------+------------+
| FxDH                | biparental | Fx, followed by DH generation                              | 8          |
+---------------------+------------+------------------------------------------------------------+------------+
| BCx                 | biparental | backcross, second parent is recurrent parent               | 9          |
+---------------------+------------+------------------------------------------------------------+------------+
| BCxDH               | biparental | BCx, followed by DH generation                             | 9          |
+---------------------+------------+------------------------------------------------------------+------------+
| BC1Sx               | biparental | BC1, followed by x generations of selfing                  | 7          |
+---------------------+------------+------------------------------------------------------------+------------+
| BC1SxDH             | biparental | BC1, followed by x generations of selfing and DH           | 6          |
+---------------------+------------+------------------------------------------------------------+------------+
| C3                  | three-way  | three way cross: (AxB) x C                                 |            |
+---------------------+------------+------------------------------------------------------------+------------+
| C3DH                | three-way  | C3, followed by DH generation                              |            |
+---------------------+------------+------------------------------------------------------------+------------+
| C3Sx                | three-way  | C3, followed by x generations of selfing                   | 7          |
+---------------------+------------+------------------------------------------------------------+------------+
| C3SxDH              | three-way  | C3, followed by x generations of selfing and DH generation | 6          |
+---------------------+------------+------------------------------------------------------------+------------+
| C4                  | four-way   | four-way cross: (AxB) x (CxD)                              |            |
+---------------------+------------+------------------------------------------------------------+------------+
| C4DH                | four-way   | C4, followed by DH generation                              |            |
+---------------------+------------+------------------------------------------------------------+------------+
| C4Sx                | four-way   | C4, followed by x generations of selfing                   | 6          |
+---------------------+------------+------------------------------------------------------------+------------+
| C4SxDH              | four-way   | C4, followed by x generations of selfing and DH generation | 6          |
+---------------------+------------+------------------------------------------------------------+------------+

### Example IBD Calculations

As an example for performing IBD calculations for a multi-parent population we use data from a maize NAM population described in @Giraud2014. The NAM population consists of 10 bi-parental doubled haploid (DH) crosses with central parent F353. The total population consists of 841 individuals. Several traits were measured in four locations across Europe. In each location an incomplete block design was used. We calculated the best linear unbiased estimations (BLUEs) of those traits using the R package [statgenSTA](https://biometris.github.io/statgenSTA). As an example, we perform QTL mapping for only the mean value of the number of days to silking ("mean_DtSILK") across all locations. The data for this population is available from the package in zipped format.

Before doing QTL detection we compute IBD probabilities on a grid of positions along the genome. This can be done using the `calcIBDmpp` function in the `statgenMPP` package. To perform IBD calculations a cross file is required for each of the populations. These files should be a tab-delimited file with first column ID identifying the genotype. The following columns should contain marker information. The first rows should contain the parents used in the cross. As an example, the file for the first cross starts like this:

| ID | PZE.101000088 | PZE.101000344 | PZE.101000449 | ... |
|---|---|---|---|---|
| F353 | 2 | 1 | 1 | ... |
| B73 | 2 | 2 | 1 | ... |
| CFD02-003 | 2 | 2 | 1 | ... |
| CFD02-006 | 2 | 2 | 1 | ... |

A map file is also required. This should also be a tab-delimited file with three columns, "marker name", "chromosome" and "position". The map file has to be identical for all crosses.

Phenotypic data can be added as a `data.frame` when computing IBD probabilities. Such a `data.frame` should have a first column "genotype" and all other columns have to be numerical. For the maize NAM population the `data.frame` with phenotypic data starts like this:

| genotype | INR_DMY | KWS_DMY | Syngenta_DMY | TUM_DMY | ... | 
|---|---|---|---|---|---|
| CFD02-003 | 182.34 | NA | 173.37 | NA | ... | 
| CFD02-006 | 171.95 | 200.56 | 159.29 | 243.29 | ... | 
| CFD02-010 | 207.56 | 236.54 | 159.38 | 228.14 | ... | 
| CFD02-024 | 185.10 | 218.94 | 160.72 | 214.44 | ... | 
| CFD02-027 | 206.46 | 226.25 | 185.24 | 221.02 | ... | 
| CFD02-036 | 180.28 | 223.24 | 162.68 | 215.86 | ... | 

In total there are 30 columns with different traits. When performing IBD calculations and specifying phenotypic data, the phenotypic data will be combined with computed IBD probabilities based. For this the genotype specified in the ID column in the cross file(s) will be matched with the genotype in the genotype column in the `data.frame` with phenotypic information. Phenotypic data for all crosses can either be added from a single `data.frame` containing phenotypic data for all crosses, or a `list` of `data.frames` each containing phenotypic data for a single cross.

```{r maizeIBD}
## Define names of crosses.
crosses <- paste0("F353x", c("B73", "D06", "D09", "EC169", "F252", "F618", 
                             "Mo17", "UH250", "UH304", "W117"))
head(crosses)

## Specify files containing crosses.
## Extract them in a temporary directory.
tempDir <- tempdir()
crossFiles <- unzip(system.file("extdata/maize/maize.zip", package = "statgenMPP"), 
                    files = paste0(crosses, ".txt"), exdir = tempDir)

## Specify file containing map.
mapFile <- unzip(system.file("extdata/maize/maize.zip", package = "statgenMPP"), 
                 files = "map.txt", exdir = tempDir)

## Read phenotypic data.
phenoFile <- unzip(system.file("extdata/maize/maize.zip", package = "statgenMPP"),
                   files = "EUmaizePheno.txt", exdir = tempDir)
phenoDat <- read.delim(phenoFile)
head(phenoDat[, 1:5])

## Perform IBD calculations. 
maizeMPP <- calcIBDmpp(crossNames = crosses, 
                       markerFiles = crossFiles,
                       pheno = phenoDat,
                       popType = "DH",
                       mapFile = mapFile,
                       evalDist = 5)
```

With `calcIBDmpp` IBD probabilities are computed on a grid for each of the crosses separately and the combined into a single output object. The value of `evalDist` can be used to specify the maximum distance between two evaluation points on the grid. The exact distance depends on the length of the chromosomes. The output is stored in an object of class `gData` (genomic Data) in which information about map, markers, and phenotypic data is combined. With the `summary` function we can get some insight in this information.

```{r sumMaizeIBD}
## Print summary
summary(maizeMPP, traits = c("INR_DMY", "KWS_DMY", "Syngenta_DMY", 
                             "TUM_DMY", "mean_DMY"))

```

To get a further idea about the population and the computed IBD probabilities we can visualize the results. First we have a look at structure of the pedigree of the population using `plotType = "pedigree"` to get a genereal idea of what the design looks like.

```{r plotPmaizeIBD}
## Plot structure of the pedigree.
plot(maizeMPP, plotType = "pedigree")
```

Next we look at the genetic map using `plotType = "genMap"`. This will display the genetic map of the population showing the length of each of the chromosomes and indicating the positions where the IBD probabilities were calculated. Optionally it is possible to highlight one or more markers using `highlight` argument.

```{r plotGMmaizeIBD}
## Plot genetic map.
# Highlight marker on chromosome 3 at position 121.61.
plot(maizeMPP, plotType = "genMap", highlight = "EXT_3_121.61")
```

Finally we visualize the computed IBD probabilities across the genome for a selected genotype using `plotType = "singleGeno"`. This plot will show the IBD probabilities for all parents for all positions on the genome for the selected genotype.

```{r plotSGmaizeIBD, fig.height=10}
## Plot IBD probabilities for genotype CFD02-003.
plot(maizeMPP, plotType = "singleGeno", genotype = "CFD02-003")
```

### Example RABBIT

Instead of performing IBD calculations directly with the package, it is also possible to import IBD probabilities computed using RABBIT. The main advantage of using RABBIT for IBD calculations is that it can handle complex pedigree populations and therefore can also be used in cases where the population structure is more complex than those described in the table above, e.g. in the NAM population described before.

As an example we use a barley population described in @Liller2017-jz. This MPP design consists of 5 parents. Four wild parents were crossed with the cultivar Morex and then backcrossed with Morex once. Individuals from the four families from the backcrosses were then crossed with each other as in a full diallel design, which generated six F6 families through five generations of selfing. The trait of interest for this population is awn length ("Awn_length"). As for the NAM population, the data for this population is available in zipped format in the package.

RABBIT output can be read using the `readRABBIT` function in `statgenMPP`. This has as input the standard RABBIT output summary file and the pedigree file that needs to be provided to RABBIT as well. This pedigree file is an optional input and is only used for plotting the pedigree structure of the population. Without it QTL mapping can still be performed. As for `calcIBDmpp` the phenotypic data has to be provided as a `data.frame`. This `data.frame` has been included in the package. 

```{r barleyIBD}
## Specify files containing RABBIT output.
## Extract in a temporary directory.
tempDir <- tempdir()
inFile <- unzip(system.file("extdata/barley/barley_magicReconstruct.zip", 
                            package = "statgenMPP"), exdir = tempDir)

## Specify pedigree file.
pedFile <- system.file("extdata/barley/barley_pedInfo.csv",
                       package = "statgenMPP")

## Read phenotypic data.
data("barleyPheno")

## read RABBIT output. 
barleyMPP <- readRABBIT(infile = inFile,
                        pedFile = pedFile,
                        pheno = barleyPheno)
```

As for the maize example we can plot the imported data to get a first idea of its content.

```{r plotPbarleyIBD}
## Plot structure of the pedigree.
plot(barleyMPP, plotType = "pedigree")
```

```{r plotGMbarleyIBD}
## Plot genetic map.
plot(barleyMPP, plotType = "genMap")
```

```{r plotAGbarleyIBD}
## Plot IBD probabilities for genotype AB-13-1-1-2
plot(barleyMPP, plotType = "singleGeno", genotype = "AB-13-1-1-2")
```

## QTL Mapping

The actual QTL mapping can be done in two steps. First we can do a Simple Interval Mapping (SIM) to get a global idea of the QTL profile. After that Composite Interval Mapping (CIM) can be done.     

We follow the methods described in @Li2021 for QTL mapping. In the first step a model without markers is fitted:

$$y = X\beta + \varepsilon, \quad \varepsilon \sim { }N\left( {0, \oplus_{k = 1}^{F} I_{{n_{k} }} \sigma_{{\varepsilon_{k} }}^{2} } \right),$$
where $y$ is a vector with phenotypes, $X$ the design matrix for the cross structure, $\beta$ a vector of fixed cross intercepts, and the residual term $\varepsilon$ has a cross-specific variance covariance structure  $\oplus_{k = 1}^{F} I_{{n_{k} }}$ in which $\sigma_{{\varepsilon_{k} }}^{2} $ is the residual variance of the $k^{th}$ cross with family size $n_{k}$. For an n-way cross where $k = 1$, e.g. MAGIC designs, the variance covariance structure of residual term is homogeneous.

Then for each of the evaluation points $q$ in the map a second model is fitted:

$$y = X\beta + M_{q} a_{q} + \varepsilon, a_{q} \sim { }N\left( {0,I_{P} \sigma_{q}^{2} } \right), \varepsilon \sim { }N\left( {0, \oplus_{k = 1}^{F} I_{{n_{k} }} \sigma_{{\varepsilon_{k} }}^{2} } \right),$$
where $M_{q}$ is the design matrix containing the expected number of parental alleles as a function of IBD probabilities, $a_{q}$ is the vector of random parental effects with variance covariance structure $I_{P} \sigma_{q}^{2}$, where $\sigma_{q}^{2}$ is the genetic variance of the QTL effects at P parents.

Variance components of a putative QTL from the two models are evaluated to perform the likelihood ratio test (LRT). The p-value is calculated using a $\chi^2$ mixture approximation for the Likelihood Ratio Test. 

For CIM the first steps are the same. However after the first round of scanning the QTL with the highest $-10log(p)$ value is added as a cofactor in the model and a new round of scanning is done. This is repeated until no new QTLs are found with a $-10log(p)$ value above the predefined `threshold`, or until the predefined maximum number of cofactors (`maxCofactors`) is added to the model. In a region of width `QTLwindow` around a cofactor no new cofactors can be detected.

#### Example maize

Going back to our maize example we can first perform SIM.

```{r saveMaize, include=FALSE}
## Use saved data to speed up vignette building.
#usethis::use_data(maizeSIM, maizeCIM, internal = FALSE, overwrite = TRUE)
```

```{r maizeSIM, eval=FALSE}
## Perform simple interval mapping.
maizeSIM <- selQTLmpp(MPPobj = maizeMPP,
                      trait = "mean_DtSILK",
                      maxCofactors = 0)
```

The results of the SIM can be plotted using `plot` function with `plotType = "QTLProfile"`. 

```{r QPmaizeSIM}
## Plot QTL Profile for maize SIM.
plot(maizeSIM, plotType = "QTLProfile")
```

Now we can do CIM. Based on the profile plot for the SIM the threshold is set to 5 to restrict a bit the number of QTLs that will be detected.

```{r maizeCIM, eval=FALSE}
## Perform composite interval mapping.
maizeCIM <- selQTLmpp(MPPobj = maizeMPP,
                      trait = "mean_DtSILK",
                      threshold = 5)
```

Now we can plot the positions of the QTLs found on the genetic map. This will produce a plot similar the the genetic map plot we have seen before, but now the QTLs will be highlighted. This plot can be made by specifying `plotType = "QTLRegion"`

```{r plotQRmaizeCIM}
##Plot QTL Profile for maize CIM.
plot(maizeCIM, plotType = "QTLRegion")
```

As for the SIM we can also plot the QTL profile. The QTLs found will be highlighted in the profile in red.

```{r plotQPmaizeCIM}
##Plot QTL Profile for maize CIM.
plot(maizeCIM, plotType = "QTLProfile")
```

It is also possible to plot the size of the parental effects for each of the QTLs found. Positive effects of a parent on the trait will be indicated by shades of red, negative effects by shades of blue. The stronger the color, the larger the effect for the specific parent is. This plot can be made using `plotType = "parEffs"`.

```{r plotPEmaizeCIM}
##Plot QTL Profile for maize CIM.
plot(maizeCIM, plotType = "parEffs")
```

Finally a combined plot of the QTL profile and the parental can be made. In this plot the two previous plots are plotted above each other with the chromosomes and positions aligned to allow for easily getting an overview of which effect belongs to which QTL in the QTL profile. This plot can be made using `plotType = "QTLProfileExt"`.

```{r plotQPEmaizeCIM}
##Plot QTL Profile for maize CIM.
plot(maizeCIM, plotType = "QTLProfileExt")
```

From the output of `selQTLmpp` the p-Values and effects for all markers can be extracted. They are stored in a `data.table` within the output object. The example below shows how to extract them. The output will contain columns snp, chr and pos with name, chromosome number and position of the marker and columns pValue and eff_par1, eff_par2, eff_par.. with the effects of all parents for that marker.

```{r extractRes}
## Extract results of QTL mapping.
maizeCIMres <- maizeCIM$GWAResult$pheno
head(maizeCIMres[, 1:8])
```
It is also possible to only extract the markers that are either QTLs or within the window of one of the selected QTLs, as specified when calling the `selQTLmpp` function. As for the full results, this information in stored in the output as a `data.table` that can be extracted as shown below. The columns in this `data.table` are identical to those in the full results except for an additional column at the end, snpStatus, that shows whether a marker is a QTL or within the window of a QTL.

```{r extractQTL}
## Extract QTLs and markers within QTL windows.
maizeCIMQTL <- maizeCIM$signSnp$pheno
head(maizeCIMQTL[, c(1:8, 18)])
```

```{r winddown, include = FALSE}
options(op)
```

------------------------------------------------------------------------

## References

