<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Multi round genome scans for QTL detection — selQTLMPP • statgenMPP</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Multi round genome scans for QTL detection — selQTLMPP"><meta name="description" content="Multi round genome scans for QTL detection.
Several rounds of QTL detection are performed. First a model is fitted
without cofactors. If for at least one marker the \(-log10(p)\) value is
above the threshold the marker with the lowest p-Value is added as cofactor
in the next round of QTL detection. This process continues until there are
no new markers with a \(-log10(p)\) value above the threshold or until
the maximum number of cofactors is reached."><meta property="og:description" content="Multi round genome scans for QTL detection.
Several rounds of QTL detection are performed. First a model is fitted
without cofactors. If for at least one marker the \(-log10(p)\) value is
above the threshold the marker with the lowest p-Value is added as cofactor
in the next round of QTL detection. This process continues until there are
no new markers with a \(-log10(p)\) value above the threshold or until
the maximum number of cofactors is reached."><meta property="og:image" content="/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">statgenMPP</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/index.html">Vignettes</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="../news/index.html"><span class="fa fa-newspaper-o"></span> news</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-file-code-o"></span> functions</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-other-packages" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true"><span class="fa fa-seedling"></span> other packages</button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-other-packages"><li><a class="external-link dropdown-item" href="https://biometris.github.io/statgenSTA">statgenSTA</a></li>
    <li><a class="external-link dropdown-item" href="https://biometris.github.io/statgenGxE">statgenGxE</a></li>
    <li><a class="external-link dropdown-item" href="https://biometris.github.io/statgenGWAS">statgenGWAS</a></li>
    <li><a class="external-link dropdown-item" href="https://biometris.github.io/statgenQTLxT">statgenQTLxT</a></li>
    <li><a class="external-link dropdown-item" href="https://biometris.github.io/statgenHTP">statgenHTP</a></li>
    <li><a class="external-link dropdown-item" href="https://biometris.github.io/statgenIBD">statgenIBD</a></li>
    <li><a class="external-link dropdown-item" href="https://biometris.github.io/LMMsolver">LMMsolver</a></li>
  </ul></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Biometris/statgenMPP"><span class="fa fa-github fa-lg fab"></span> github</a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Multi round genome scans for QTL detection</h1>

      <div class="d-none name"><code>selQTLMPP.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Multi round genome scans for QTL detection.<br><br>
Several rounds of QTL detection are performed. First a model is fitted
without cofactors. If for at least one marker the \(-log10(p)\) value is
above the threshold the marker with the lowest p-Value is added as cofactor
in the next round of QTL detection. This process continues until there are
no new markers with a \(-log10(p)\) value above the threshold or until
the maximum number of cofactors is reached.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">selQTLMPP</span><span class="op">(</span></span>
<span>  <span class="va">MPPobj</span>,</span>
<span>  trait <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  QTLwindow <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  maxCofactors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  K <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  computeKin <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-mppobj">MPPobj<a class="anchor" aria-label="anchor" href="#arg-mppobj"></a></dt>
<dd><p>An object of class gDataMPP, typically the output of either
<code><a href="calcIBDmpp.html">calcIBDMPP</a></code> or <code><a href="readRABBITMPP.html">readRABBITMPP</a></code>.</p></dd>


<dt id="arg-trait">trait<a class="anchor" aria-label="anchor" href="#arg-trait"></a></dt>
<dd><p>A character string indicating the trait QTL mapping is done for.</p></dd>


<dt id="arg-qtlwindow">QTLwindow<a class="anchor" aria-label="anchor" href="#arg-qtlwindow"></a></dt>
<dd><p>A numerical value indicating the window around a QTL that
is considered as part of that QTL.</p></dd>


<dt id="arg-threshold">threshold<a class="anchor" aria-label="anchor" href="#arg-threshold"></a></dt>
<dd><p>A numerical value indicating the threshold for the
\(-log10p\) value of a marker to be considered a QTL.</p></dd>


<dt id="arg-maxcofactors">maxCofactors<a class="anchor" aria-label="anchor" href="#arg-maxcofactors"></a></dt>
<dd><p>A numerical value, the maximum number of cofactors to
include in the model. If <code>NULL</code> cofactors are added until no new
cofactors are found.</p></dd>


<dt id="arg-k">K<a class="anchor" aria-label="anchor" href="#arg-k"></a></dt>
<dd><p>A list of chromosome specific kinship matrices. If
<code>NULL</code> and <code>computeKin = FALSE</code> no kinship matrix is included in
the models.</p></dd>


<dt id="arg-computekin">computeKin<a class="anchor" aria-label="anchor" href="#arg-computekin"></a></dt>
<dd><p>Should chromosome specific kinship matrices be computed?</p></dd>


<dt id="arg-parallel">parallel<a class="anchor" aria-label="anchor" href="#arg-parallel"></a></dt>
<dd><p>Should the computation of variance components be done in
parallel? This requires a parallel back-end to be registered. See examples.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Should progress and intermediate plots be output?</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An object of class <code>QTLMPP</code></p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>By default only family specific effects and residual variances and no
kinship relations are included in the model. It is possible to include
kinship relations by either specifying <code>computeKin = TRUE</code>. When doing
so the kinship matrix is computed by averaging \(Z Z^t\) over all markers,
where \(Z\) is the genotype x parents matrix for the marker. It is also
possible to specify a list of precomputed chromosome
specific kinship matrices in <code>K</code>. Note that adding a kinship matrix
to the model increases the computation time a lot, especially for large
populations.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="kinshipIBD.html">kinshipIBD</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Read phenotypic data.</span></span></span>
<span class="r-in"><span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/multipop"</span>, <span class="st">"AxBxCpheno.txt"</span>,</span></span>
<span class="r-in"><span>                               package <span class="op">=</span> <span class="st">"statgenMPP"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">## Rename first column to genotype.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"genotype"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Compute IBD probabilities for simulated population - AxB, AxC.</span></span></span>
<span class="r-in"><span><span class="va">ABC</span> <span class="op">&lt;-</span> <span class="fu"><a href="calcIBDmpp.html">calcIBDMPP</a></span><span class="op">(</span>crossNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AxB"</span>, <span class="st">"AxC"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                  markerFiles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/multipop"</span>, <span class="st">"AxB.txt"</span>,</span></span>
<span class="r-in"><span>                                              package <span class="op">=</span> <span class="st">"statgenMPP"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/multipop"</span>, <span class="st">"AxC.txt"</span>,</span></span>
<span class="r-in"><span>                                              package <span class="op">=</span> <span class="st">"statgenMPP"</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                  pheno <span class="op">=</span> <span class="va">pheno</span>,</span></span>
<span class="r-in"><span>                  popType <span class="op">=</span> <span class="st">"F4DH"</span>,</span></span>
<span class="r-in"><span>                  mapFile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/multipop"</span>, <span class="st">"mapfile.txt"</span>,</span></span>
<span class="r-in"><span>                                        package <span class="op">=</span> <span class="st">"statgenMPP"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                  evalDist <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Single-QTL Mapping.</span></span></span>
<span class="r-in"><span><span class="va">ABC_SQM</span> <span class="op">&lt;-</span> <span class="fu">selQTLMPP</span><span class="op">(</span><span class="va">ABC</span>, trait <span class="op">=</span> <span class="st">"yield"</span>, maxCofactors <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Multi-QTL Mapping.</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co">## Register parallel back-end with 2 cores.</span></span></span>
<span class="r-in"><span><span class="fu">doParallel</span><span class="fu">::</span><span class="fu">registerDoParallel</span><span class="op">(</span>cores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Run multi-QTL mapping.</span></span></span>
<span class="r-in"><span><span class="va">ABC_MQM</span> <span class="op">&lt;-</span> <span class="fu">selQTLMPP</span><span class="op">(</span><span class="va">ABC</span>, trait <span class="op">=</span> <span class="st">"yield"</span>, parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Run multi-QTL mapping - include kinship matrix.</span></span></span>
<span class="r-in"><span><span class="va">ABC_MQM_kin</span> <span class="op">&lt;-</span> <span class="fu">selQTLMPP</span><span class="op">(</span><span class="va">ABC</span>, trait <span class="op">=</span> <span class="st">"yield"</span>, parallel <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                        computeKin <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Martin Boer, Wenhao Li, Bart-Jan van Rossum.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

